bittensor/_axon/__init__.py

```diff
@@ -441,19 +453,21 @@ def check_signature(
             message = f"{nonce}{sender_hotkey}{receptor_uuid}"
         else:
             raise Exception("Invalid signature version")
-        # Build the key which uniquely identifies the endpoint that has signed
-        # the message.
-        endpoint_key = f"{sender_hotkey}:{receptor_uuid}"
 
         if endpoint_key in self.nonces.keys():
             previous_nonce = self.nonces[endpoint_key]
             # Nonces must be strictly monotonic over time.
-            if nonce <= previous_nonce:
-                raise Exception("Nonce is too small")
+            lower_bound = clock.monotonic_ns() - bittensor.__blocktime__*1e9 # nonce should not be too much earlier then block_time
+            upper_bound = clock.monotonic_ns() + bittensor.__blocktime__*1e9 # nonce should not be too much older then block_time
+            if nonce < lower_bound or nonce > upper_bound or nonce <= previous_nonce:
+                raise Exception(f"Nonce {nonce} is out of range {lower_bound} to {upper_bound} OR nonce is earlier then previous nonce {previous_nonce}.")
 
         if not keypair.verify(message, signature):
             raise Exception("Signature mismatch")
-        self.nonces[endpoint_key] = nonce
+
+        ss58_decode(sender_hotkey) # verify it is a valid hotkey to be written, otherwise raise value error
```

Comment:
Why is this needed ? The initializer for `Keypair`, on line 448, already does this:
https://github.com/polkascan/py-substrate-interface/blob/1f558e6a5fa9a1ea71ae25211989ebf858a14075/substrateinterface/keypair.py#L108-L109